src/snabbdom.ts

```diff
@@ -60,125 +95,111 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
     }
   }
 
-  function emptyNodeAt(elm: Element) {
-    const id = elm.id ? '#' + elm.id : '';
-    const c = elm.className ? '.' + elm.className.split(' ').join('.') : '';
-    return vnode(api.tagName(elm).toLowerCase() + id + c, {}, [], undefined, elm);
-  }
-
   function createRmCb(childElm: Node, listeners: number) {
     return function rmCb() {
       if (--listeners === 0) {
-        const parent = api.parentNode(childElm);
+        const parent = api.parentNode(childElm) as Node;
         api.removeChild(parent, childElm);
       }
     };
   }
 
   function createElm(vnode: VNode, insertedVnodeQueue: VNodeQueue): Node {
-    let i: any, data = vnode.data;
-    if (data !== undefined) {
+    let i: any;
+    const {data} = vnode;
+    if (!isUndef(data)) {
       if (isDef(i = data.hook) && isDef(i = i.init)) {
         i(vnode);
-        data = vnode.data;
       }
     }
-    let children = vnode.children, sel = vnode.sel;
-    if (sel === '!') {
-      if (isUndef(vnode.text)) {
-        vnode.text = '';
-      }
-      vnode.elm = api.createComment(vnode.text as string);
-    } else if (sel !== undefined) {
-      // Parse selector
-      const hashIdx = sel.indexOf('#');
-      const dotIdx = sel.indexOf('.', hashIdx);
-      const hash = hashIdx > 0 ? hashIdx : sel.length;
-      const dot = dotIdx > 0 ? dotIdx : sel.length;
-      const tag = hashIdx !== -1 || dotIdx !== -1 ? sel.slice(0, Math.min(hash, dot)) : sel;
-      const elm = vnode.elm = isDef(data) && isDef(i = (data as VNodeData).ns) ? api.createElementNS(i, tag)
-                                                                               : api.createElement(tag);
-      if (hash < dot) elm.setAttribute('id', sel.slice(hash + 1, dot));
-      if (dotIdx > 0) elm.setAttribute('class', sel.slice(dot + 1).replace(/\./g, ' '));
+    if (isElementVNode(vnode)) {
+      const {data, tag} = vnode;
+      const elm = vnode.elm = isDef(i = data.ns) ? api.createElementNS(i, tag)
+                                                 : api.createElement(tag);
+      if (isDef(i = data.hook) && isDef(i.create)) i.create(emptyNode, vnode);
```

Comment:
this is another important change that requires more debate. In the previous iteration, the `data.hook.create` happens after the children is processed, while the `module.create` happens before. In this refactor, they are unified, and are happening at the same time. I was confused on the "why", could not really explain it.

Additionally, the more I understand the code, the more I think that `data.hook.*` are just redundant, and that you could create a module that emulates those hooks instead of having snabbdom to take care of it. Maybe that's a better option, to implement a optional module called hooks, that provides the exact same semantics that we have today, and you can use it if you really want to explore that avenue. What do you think?